langchain4j-mcp/src/main/java/dev/langchain4j/mcp/McpToolProvider.java

```diff
@@ -0,0 +1,52 @@
+package dev.langchain4j.mcp;
+
+import dev.langchain4j.agent.tool.ToolSpecification;
+import dev.langchain4j.mcp.client.McpClient;
+import dev.langchain4j.service.tool.ToolProvider;
+import dev.langchain4j.service.tool.ToolProviderRequest;
+import dev.langchain4j.service.tool.ToolProviderResult;
+import java.util.List;
+
+/**
+ * A tool provider backed by one or more MCP clients
+ */
+public class McpToolProvider implements ToolProvider {
+
+    private final List<McpClient> mcpClients;
+
+    private McpToolProvider(Builder builder) {
+        this.mcpClients = builder.mcpClients;
```

Comment:
copy the list?

Reply:
I've now changed it to wrap it in `Collections.unmodifiableList`, do we need a full copy?!

Reply:
By full do you mean deep? I am not sure. In theory user can manipulate `McpClient`: call `listTools`/`executeTool`/`close`. Can this interfere if the same `McpClient` will be used simultaneously by user and `McpToolProvider`? I guess it can, but should we care?

Reply:
Ok I'll do `this.mcpClients = new ArrayList<>(builder.mcpClients);` here just to be sure